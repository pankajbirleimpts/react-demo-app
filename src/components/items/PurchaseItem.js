import React, { Component } from 'react';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import { Grid } from 'semantic-ui-react';
import moment from 'moment';
import { toast } from 'react-toastify';
import {
  getAllItems,
  purchaseItem,
  getDayItem,
  getAllDayItems
} from '../../actions/ItemAction';
import { getAllUsers } from '../../actions/UserAction';
import { langs } from '../../config';
import { Loader } from '../common';
import './dayitem.css';

export class UnconnectedPurchaseItem extends Component {
  state = {
    id: '',
    otherTransaction: false,
    initialValues: {
      itemId: '',
      userId: '',
      userDetails: null,
      itemDetails: null,
      date: '',
      quantity: '',
      amount: 0
    }
  };

  componentDidMount() {
    // Get all items
    const todayDate = moment().format('DD-MMM-YYYY');
    this.props.getAllDayItems(todayDate);
    // get all users
    this.props.getAllUsers();
  }
  /**
   * @formSubmitHandler
   * @desc: Submit the signup form
   */
  formSubmitHandler = values => {
    const dayItem = this.props.item.allDayItems.find(
      val => val.id === values.itemId
    );
    const userDetails = this.props.user.allUsers.find(
      val => val.id === values.userId
    );
    const purchaseAmount = this.getTotalAmount(values.itemId, values.quantity);
    // If user do not have sufficient balance
    if (this.getUserBalance(userDetails.id) < purchaseAmount) {
      toast.warn(langs.messages.INSUFFICIENT_BALANCE);
      return false;
    }

    const remainingQuantity = dayItem.remainingQuantity + values.quantity;
    if (remainingQuantity > dayItem.totalQuantity) {
      toast.warn(langs.messages.INSUFFICIENT_QUANTITY);
      return false;
    }
    // Update remaining quantity
    dayItem.remainingQuantity = remainingQuantity;
    // Update user balance
    userDetails.balance = userDetails.balance - purchaseAmount;

    values.userDetails = userDetails;
    values.itemDetails = dayItem.itemDetails;
    values.date = moment().format('DD-MMM-YYYY');
    values.purchaseAmount = purchaseAmount;

    this.props.purchaseItem(userDetails, values, dayItem, () => {
      const state = this.state;
      if (state.otherTransaction) {
        const initialValues = {
          ...this.state.initialValues,
          userId: values.userId,
          amount: ''
        };
        this.setState({ initialValues, otherTransaction: false });
      } else {
        this.setState({
          otherTransaction: false,
          initialValues: {
            ...this.state.initialValues,
            userId: '',
            amount: '',
            itemId: ''
          }
        });
      }
    });
  };

  /**
   * @formValidation
   * @desc: form validation
   */
  formValidation = () => {
    return Yup.object().shape({
      userId: Yup.string().required(langs.messages.REQUIRED),
      itemId: Yup.string().required(langs.messages.REQUIRED),
      quantity: Yup.number()
        .required(langs.messages.REQUIRED)
        .test(
          'quantityValidation',
          langs.messages.QUANTITY_VALIDATION,
          function(value) {
            if (value > 0) {
              return true;
            } else {
              return false;
            }
          }
        )
    });
  };

  // Get item amount
  getItemAmount = itemId => {
    if (itemId) {
      const selectedItem = this.props.item.allDayItems.find(
        val => val.id === itemId
      );
      if (selectedItem) {
        return selectedItem.amount;
      }
      return 0;
    }
    return 0;
  };

  // Get item remaining quantity
  getItemRemainQuantity = itemId => {
    if (itemId) {
      const selectedItem = this.props.item.allDayItems.find(
        val => val.id === itemId
      );
      if (selectedItem) {
        return selectedItem.totalQuantity - selectedItem.remainingQuantity;
      }
      return 0;
    }
    return 0;
  };

  //Get total amount
  getTotalAmount = (itemId, quantity) => {
    const itemPrice = this.getItemAmount(itemId);
    return (itemPrice * quantity).toFixed(2);
  };

  handleClickMakeOtherTransaction = () => {
    this.setState({
      otherTransaction: true
    });
  };

  getUserBalance = userId => {
    if (userId) {
      const userDetails = this.props.user.allUsers.find(
        user => user.id === userId
      );
      if (userDetails) return userDetails.balance;
      return 0;
    }
    return 0;
  };

  /**
   * @renderForm
   * @desc: rednder form
   */
  renderForm = ({ values }) => {
    return (
      <Form noValidate className='ui form'>
        <div className='field'>
          <label>User *</label>
          <Field name='userId' as='select' data-test="userId-input">
            <option value=''>Please Select</option>
            {this.props.user.allUsers
              .filter(val => val.role === 'EMPLOYEE')
              .map(val => {
                return (
                  <option key={val.id} value={val.id}>
                    {val.firstName} {val.lastName}
                  </option>
                );
              })}
          </Field>
          {values.userId && (
            <div className='show-user-balance'>
              Current Balance: Rs. {this.getUserBalance(values.userId)}
            </div>
          )}
          <ErrorMessage component='p' name='userId' className='red' />
        </div>
        <div className='field'>
          <label>Item *</label>
          <Field name='itemId' as='select' data-test="itemId-input">
            <option value=''>Please Select</option>
            {this.props.item.allDayItems.map(val => {
              return (
                <option key={val.id} value={val.id}>
                  {val.itemName}
                </option>
              );
            })}
          </Field>
          <ErrorMessage component='p' name='itemId' className='red' />
          {this.getItemAmount(values.itemId) > 0 && (
            <div className='show-amount'>
              Rs. {this.getItemAmount(values.itemId)} {'   '}(Remaining Qty:{' '}
              {this.getItemRemainQuantity(values.itemId)})
            </div>
          )}
        </div>
        <div className='field'>
          <label>Quantity *</label>
          <Field name='quantity' type='number' data-test="quantity-input"/>
          <ErrorMessage component='p' name='quantity' className='red' />
          {this.getTotalAmount(values.itemId, values.quantity) > 0 && (
            <div className='show-total-amount'>
              Total: Rs. {this.getTotalAmount(values.itemId, values.quantity)}
            </div>
          )}
        </div>        
        <button className='ui button primary' type='submit'>
          Submit
        </button>
        <button
          data-test="other-transation-button"
          className='ui button primary'
          type='submit'
          onClick={() => {
            this.handleClickMakeOtherTransaction();
          }}
        >
          Submit & Make Other Transaction
        </button>
      </Form>
    );
  };

  render() {
    return (
      <Grid>
        <Loader isLoading={this.props.item.isLoading} />
        <Grid.Row centered className='add-day-item-container'>
          <Grid.Column width='8'>
            <h2 data-test="page-heading">Purchase Item</h2>
            <Formik
              enableReinitialize
              initialValues={this.state.initialValues}
              validationSchema={this.formValidation()}
              onSubmit={values => {
                this.formSubmitHandler(values);
              }}
            >
              {this.renderForm}
            </Formik>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}

function mapStateToProp({ item, user }) {
  return {
    item,
    user
  };
}

export default connect(mapStateToProp, {
  purchaseItem,
  getDayItem,
  getAllItems,
  getAllUsers,
  getAllDayItems
})(UnconnectedPurchaseItem);
