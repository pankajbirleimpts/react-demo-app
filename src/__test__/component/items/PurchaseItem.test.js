import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { BrowserRouter } from 'react-router-dom';
import PurchaseItem, { UnconnectedPurchaseItem } from '../../../components/items/PurchaseItem';
import { storeFactory, findByAttr } from '../../test/TestUtils';


const setup = (props) => {
    const wrapper = shallow(
        <UnconnectedPurchaseItem {...props} />,
    )
    return wrapper;
};

const setupWithStore = (props) => {
    const store = storeFactory(props);
    const wrapper = shallow(
        <PurchaseItem store={store} />,
    ).dive().dive();
    return wrapper;
};

describe('Purchase Item Flow', () => {
    const props = {
        item: {
            isLoading: false,
            allDayItems: [{
                id: 1,
                date: "19-Feb-2020",
                itemId: 101,
                remainingQuantity: 0,
                totalQuantity: 5,
                itemName: "Pahe",
                itemDetails: {
                    id: 101,
                    itemName: "Pahe",
                    category: "Breakfast",
                }
            }]
        },
        user: {
            allUsers: [{
                id: 201,
                "balance": 0,
                "email": "Ram@test.com",
                "employeeId": 201,
                "firstName": "Ram",
                "lastName": "Singh",
                "role": "ADMIN"
            }]
        },
        history: {
            replace: jest.fn(),
        },
        getAllDayItems: jest.fn(),
        getAllUsers: jest.fn(),
        purchaseItem: jest.fn(),
    }

    test('Snapshot of PurchaseItem component', () => {
        const wrapper = setup(props);
        expect(wrapper).toMatchSnapshot();
    });


    test('Check the isLoading in props', () => {
        const wrapper = setup(props);
        const componentProps = wrapper.instance().props;
        expect(componentProps.item.isLoading).toBe(props.item.isLoading);
    });

    test('Check `getAllDayItems` & `getAllUsers` action should not be called props', () => {
        const wrapper = setup(props);
        wrapper.instance().componentDidMount();
        expect(props.getAllDayItems).toHaveBeenCalled();
        expect(props.getAllUsers).toHaveBeenCalled();
    });

    test('Check page titile should be `Purchase item`', () => {
        const wrapper = setup(props);
        const pageTitle = findByAttr(wrapper, 'page-heading');
        expect(pageTitle.text()).toEqual('Purchase Item');
    });

    test('Checking the form values ', async () => {
        const formValues = {
            userId: 201,
            itemId: 1,
            quantity: 2
        };

        const wrapper = mount(<BrowserRouter><UnconnectedPurchaseItem {...props} /></BrowserRouter>);

        const userIdInput = findByAttr(wrapper, 'userId-input').find('select').first();
        const itemIdInput = findByAttr(wrapper, 'itemId-input').find('select').first();
        const quantityInput = findByAttr(wrapper, 'quantity-input').find('input').first();

        await act(async () => {
            userIdInput.simulate('change', { persist: () => { }, target: { name: 'userId', value: formValues.userId } });
        });

        await act(async () => {
            itemIdInput.simulate('change', { persist: () => { }, target: { name: 'itemId', value: formValues.itemId } });
        });

        await act(async () => {
            quantityInput.simulate('change', { persist: () => { }, target: { name: 'quantity', value: formValues.quantity } });
        });

        await act(async () => {
            wrapper.find('form').simulate('submit');
        });
        expect(props.purchaseItem).toHaveBeenCalled();
    });

    test('Check `otherTransaction` state should be true if make other transaction ', () => {
        const wrapper = mount(<UnconnectedPurchaseItem {...props} />);
        const buttonElement = findByAttr(wrapper, 'other-transation-button');
        buttonElement.simulate('click');
        const otherTransactionState = wrapper.state().otherTransaction;
        expect(otherTransactionState).toBeTruthy();
    });

    test('Check `getUserBalance` function', () => {
        const wrapper = mount(<UnconnectedPurchaseItem {...props} />);
        const response = wrapper.instance().getUserBalance(12424);
        expect(response).toBeFalsy();
    });

    test('Check `getUserBalance` function without any userId', () => {
        const wrapper = mount(<UnconnectedPurchaseItem {...props} />);
        const response = wrapper.instance().getUserBalance();
        expect(response).toBeFalsy();
    });

    test('Check `getItemRemainQuantity` function', () => {
        const wrapper = mount(<UnconnectedPurchaseItem {...props} />);
        const response = wrapper.instance().getItemRemainQuantity(12424);
        expect(response).toBeFalsy();
    });

    test('Check `getItemRemainQuantity` function without any argument', () => {
        const wrapper = mount(<UnconnectedPurchaseItem {...props} />);
        const response = wrapper.instance().getItemRemainQuantity();
        expect(response).toBeFalsy();
    });

    test('Check `getUserBalance` function without any userId', () => {
        const props = {
            item: {
                isLoading: false,
                allDayItems: [{
                    id: 1,
                    date: "19-Feb-2020",
                    itemId: 101,
                    remainingQuantity: 0,
                    totalQuantity: 5,
                    itemName: "Pahe",
                    itemDetails: {
                        id: 101,
                        itemName: "Pahe",
                        category: "Breakfast",
                    }
                }]
            },
            user: {
                allUsers: [{
                    id: 201,
                    "balance": 0,
                    "email": "Ram@test.com",
                    "employeeId": 201,
                    "firstName": "Ram",
                    "lastName": "Singh",
                    "role": "ADMIN"
                }]
            }
        };
        const wrapper = setupWithStore(props);
        const componentProps = wrapper.instance().props;
        expect(componentProps.item).toBe(props.item);
    });
});
