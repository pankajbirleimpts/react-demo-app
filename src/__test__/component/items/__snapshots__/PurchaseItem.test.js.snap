// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Purchase Item Flow Snapshot of PurchaseItem component 1`] = `
<Grid>
  <Loading
    isLoading={false}
  />
  <GridRow
    centered={true}
    className="add-day-item-container"
  >
    <GridColumn
      width="8"
    >
      <h2
        data-test="page-heading"
      >
        Purchase Item
      </h2>
      <Formik
        enableReinitialize={true}
        initialValues={
          Object {
            "amount": 0,
            "date": "",
            "itemDetails": null,
            "itemId": "",
            "quantity": "",
            "userDetails": null,
            "userId": "",
          }
        }
        onSubmit={[Function]}
        validationSchema={
          ObjectSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_defaultDefault": [Function],
            "_deps": Array [],
            "_excludedEdges": Array [],
            "_exclusive": Object {},
            "_mutate": undefined,
            "_nodes": Array [
              "quantity",
              "itemId",
              "userId",
            ],
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "object",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "fields": Object {
              "itemId": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "quantity": NumberSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "quantityValidation": false,
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "number",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "userId": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
          }
        }
      >
        <Component />
      </Formik>
    </GridColumn>
  </GridRow>
</Grid>
`;
